engine/crates/runtime-local/src/rate_limiting/in_memory/key_based.rs

```diff
@@ -34,48 +37,72 @@ impl RateLimiterContext for RateLimitingContext {
     }
 }
 
-#[derive(Default)]
 pub struct InMemoryRateLimiter {
-    inner: HashMap<String, governor::DefaultKeyedRateLimiter<usize>>,
+    limiters: Arc<RwLock<HashMap<String, governor::DefaultKeyedRateLimiter<usize>>>>,
 }
 
 impl InMemoryRateLimiter {
-    pub fn runtime(config: KeyedRateLimitConfig<'_>) -> RateLimiter {
-        let mut limiter = Self::default();
+    pub fn runtime(
+        config: KeyedRateLimitConfig,
+        mut updates: mpsc::Receiver<HashMap<String, GraphRateLimit>>,
+    ) -> RateLimiter {
+        let mut limiters = HashMap::new();
 
         // add subgraph rate limiting configuration
-        for (name, rate_limit_config) in config.rate_limiting_configs {
-            limiter = limiter.with_rate_limiter(name, rate_limit_config);
+        for (name, config) in config.rate_limiting_configs {
+            let Some(limiter) = create_limiter(config) else {
+                continue;
+            };
+
+            limiters.insert(name.to_string(), limiter);
         }
 
-        RateLimiter::new(limiter)
-    }
+        let limiters = Arc::new(RwLock::new(limiters));
+        let limiters_copy = limiters.clone();
+
+        tokio::spawn(async move {
+            while let Some(updates) = updates.recv().await {
+                let mut limiters = limiters_copy.write().unwrap();
```

Comment:
:bulb: **Performance**
Using `unwrap()` on the result of `RwLock::write()` can lead to a panic if the lock is poisoned. Consider handling the error gracefully instead.
```suggestion
Replace `let mut limiters = limiters_copy.write().unwrap();` with `let mut limiters = limiters_copy.write().map_err(|_| Error::LockPoisoned)?;` to handle potential poisoning.
```